### PR TITLE
mathSqrt summary

### PR DESCRIPTION
**Pull request by mariaKt**
This PR introduces summarized rules for the `mathSqrt` function.

These rules are saving 818 steps.

I would like your opinion specifically on this rule [here](https://github.com/Pi-Squared-Inc/solidity-demo-semantics/blob/a6f66dfaa216e7101295a9e28610f425e3d8bf95/src/uniswap-summaries.md?plain=1#L3304):
```
  // Skip environment updates when not needed.
  // These occur due to the multiple block statements, introduced by the if
  // statement that the while rewrites to.
  rule <k> restoreEnv( _:Map ) ~> .Statements ~> restoreEnv( E:Map ) => restoreEnv(E) ...</k>
       <summarize> true </summarize>
       <current-function> mathSqrt </current-function> [priority(40)]
```

Due to the while statement rewriting to an if statement, as the function is executed, there are a lot of block statements that eventually lead to a sequence of
```
restoreEnv(E) ~> .Statements ~> restoreEnv(E) ~> .Statements ~> ...
```
I added this rule to skip the unneeded updates. Due to the number of while loop body executions being input dependent, I cannot specify the exact number that this is repeated, so this is a step per loop iteration.

My question is: is this rule OK to include in a summary, or is it considered a general semantic rule?

### CODE REVIEW

File: src/uniswap-summaries.md [hunk 1]
```diff
@@ -3253,6 +3253,40 @@ module SOLIDITY-UNISWAP-GETRESERVES-SUMMARY
...
+       <env> E => (y |-> var(size(STORE), uint256))
+                  (z |-> var(size(STORE) +Int 1, uint256))
+                  (x |-> var(size(STORE) +Int 2, uint256))
+       </env>
+       <store> STORE =>
+               STORE ListItem(V) // y
+                     ListItem(V) // z
+                     ListItem((V /uMInt 2p256) +MInt 1p256) // x
+        </store>
+       <current-function> FUNC => mathSqrt </current-function>
```
Comment (Robertorosmaninho) [inline src/uniswap-summaries.md:3279]: This is necessary because we're not inside the `mathSqrt` yet, right?

File: src/uniswap-summaries.md [hunk 2]
```diff
@@ -3253,6 +3253,40 @@ module SOLIDITY-UNISWAP-GETRESERVES-SUMMARY
...
+       <call-stack>... .List => ListItem(frame(K, E, FUNC)) </call-stack>
+    requires V >uMInt 3p256 [priority(40)]
+
+  rule <k> mathSqrt:Id ( v(V:MInt{256}, uint256), .TypedVals ) => v (1p256, uint256) ...</k>
```
Comment (Robertorosmaninho) [inline src/uniswap-summaries.md:3320]: Is there a reason why the cells: `<this-type>`,  `<contract-id>`, `<contract-fn-id>`, etc. wouldn't be necessary here?

File: src/uniswap-summaries.md [hunk 3]
```diff
@@ -3280,6 +3280,13 @@ module SOLIDITY-MATHSQRT-SUMMARY
        <call-stack>... .List => ListItem(frame(K, E, FUNC)) </call-stack>
     requires V >uMInt 3p256 [priority(40)]
 
+  // While loop rewrite and condition evaluation
+  rule <k> while (( x < z ) #as Cond) Body:Statement Ss:Statements => if (v(Vx <uMInt Vz, bool)) {Body while(Cond) Body} else {.Statements} ~> Ss ...</k>
+       <summarize> true </summarize>
```
Comment (Robertorosmaninho) [inline src/uniswap-summaries.md:3285]: Same question from the case `y <= 3 && y != 0` about the cell variables

**Review by Robertorosmaninho — COMMENTED** [review]
I have just a few clarification questions before approving it:

File: src/uniswap-summaries.md (hunk 1 above)
Comment (mariaKt) [reply src/uniswap-summaries.md:3279]: Yes. We are doing a call, but stopping before it returns. so are including all changes made by the call rule to the `<env>`, `<k>`, `<call-stack>` and we need to change the <current-function> as well.

Comment (Robertorosmaninho) [reply src/uniswap-summaries.md:3279]: Got it! Thanks!

File: src/uniswap-summaries.md (hunk 2 above)
Comment (mariaKt) [reply src/uniswap-summaries.md:3320]: Yes. This rule replaces the entire call. So, we do not need to try and find the code of the function, we simply return a TypedVal. We know that a `<contract-fn>` with `<contract-fn-id> mathSqrt </contract-fn-id>` exists, but we do not need to find any information from the cells in that `<contract-fn>`.

File: src/uniswap-summaries.md (hunk 3 above)
Comment (mariaKt) [reply src/uniswap-summaries.md:3285]: Here, we are already within the body of the function `mathSqrt`, i.e. to get to this rule the one doing the matching and changes to the configuration related to the call itself has already matched and done its rewriting. So, we don't need to match any of those cells.

File: src/uniswap-summaries.md (hunk 2 above)
Comment (mariaKt) [reply src/uniswap-summaries.md:3320]: You can think of it just like any other summary of en entire call (which it is with these conditions), e.g.
https://github.com/Pi-Squared-Inc/solidity-demo-semantics/blob/c94634ebc4f33417bd60dc731cee3320186be087/src/uniswap-summaries.md?plain=1#L2658

**Review by Robertorosmaninho — APPROVED** [review]
LGTM

**Pull request merged**